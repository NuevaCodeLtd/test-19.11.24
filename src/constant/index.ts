export const APP_NAME = 'Blog About English'
export const SUCCESS_STATUS = 'success' as const
export const ERROR_STATUS = 'error' as const
export const WARNING_STATUS = 'warning' as const
export const INFO_STATUS = 'info' as const
export const GENERIC_ERROR_MESSAGE = 'Something went wrong. Please try again.'
export type Language = 'en' | 'de' | 'fr'
export const AVERAGE_READING_SPEED = 200 // words per minute
export enum NavigationLinks {
  HOME = '/',
  POSTS = '/posts',
  TAGS = '/tags',
  ADD_POST = '/add-post',
  EDIT_POST = '/edit-post'
}
