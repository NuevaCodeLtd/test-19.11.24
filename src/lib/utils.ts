/* eslint-disable @typescript-eslint/no-explicit-any */
import path from 'path'
import fs from 'fs'
import { Post } from '@/common-types/types'

export function readFile(fileName: string) {
  const pathToFile = path.join(
    process.cwd(),
    `public/content/${fileName}`
  )
  const fileContent = fs.readFileSync(pathToFile, 'utf-8')

  return fileContent
}

export function writePostsToFile(posts: Post[]) {
  const filePath = path.join(process.cwd(), 'public/content/posts.json')
  fs.writeFileSync(filePath, JSON.stringify(posts, null, 2), 'utf-8')
}

export function formDataToObject(formData: FormData) {
  const obj: Record<string, any> = {}

  formData.forEach((value, key) => {
    // Handle multiple values
    if (obj[key]) {
      if (Array.isArray(obj[key])) {
        obj[key].push(value)
      } else {
        obj[key] = [obj[key], value]
      }
    } else {
      // Convert specific fields (e.g., checkboxes) from string to boolean
      if (value === 'true' || value === 'on') {
        obj[key] = true
      } else if (value === 'false') {
        obj[key] = false
      } else {
        obj[key] = value
      }
    }
  })

  return obj
}
