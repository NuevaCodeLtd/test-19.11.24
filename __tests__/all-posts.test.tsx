import { render, screen, fireEvent } from '@testing-library/react'
import AllPosts from '../src/components/all-posts'
import { useRouter, usePathname } from 'next/navigation'
import { getAllPosts } from '../src/lib/post-utils'
import {NavigationLinks} from '@/constant';

global.structuredClone = val => JSON.parse(JSON.stringify(val))

jest.mock('../src/components/posts-grid', () => {
  return jest.fn(() => <div>PostsGrid Mock</div>)
})

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
  usePathname: jest.fn()
}))

const mockPush = jest.fn()

describe('AllPosts', () => {
  const postsData = getAllPosts(1)

  beforeEach(() => {
    jest.clearAllMocks()
    ;(useRouter as jest.Mock).mockReturnValue({ push: mockPush })
    ;(usePathname as jest.Mock).mockReturnValue(NavigationLinks.POSTS)
  })

  it('should match snapshot', () => {
    const mockPostsData = structuredClone(postsData)

    const { asFragment } = render(<AllPosts postsData={mockPostsData} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders the AllPosts component with posts and pagination', () => {
    const mockPostsData = structuredClone(postsData)

    render(<AllPosts postsData={mockPostsData} />)

    expect(screen.getByText('Welcome to blog posts')).toBeInTheDocument()
    expect(screen.getByText('PostsGrid Mock')).toBeInTheDocument()
    expect(screen.getByText('Page: 1')).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /page 2/i })).toBeInTheDocument()
    expect(screen.getByRole('button', { name: /page 3/i })).toBeInTheDocument()
  })

  it('renders without pagination if there is only one page', () => {
    const mockPostsData = {
      ...structuredClone(postsData),
      totalPages: 1
    }

    render(<AllPosts postsData={mockPostsData} />)

    expect(screen.getByText('Welcome to blog posts')).toBeInTheDocument()
    expect(screen.getByText('PostsGrid Mock')).toBeInTheDocument()
    expect(screen.queryByText('Page: 1')).not.toBeInTheDocument()
    expect(screen.queryByRole('button')).not.toBeInTheDocument() // No pagination buttons
  })

  it('handles pagination change correctly', () => {
    const mockPostsData = structuredClone(postsData)

    // Act
    render(<AllPosts postsData={mockPostsData} />)

    const pageButton = screen.getByRole('button', { name: /page 2/i })
    fireEvent.click(pageButton)

    expect(mockPush).toHaveBeenCalledWith('/posts/?page=2', undefined)
  })
})