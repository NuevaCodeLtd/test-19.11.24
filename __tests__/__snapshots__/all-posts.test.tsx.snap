// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllPosts should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 css-6s8wfx-MuiTypography-root"
    >
      Welcome to my world ðŸ‘‹ ðŸŒŽ
    </h3>
    <div>
      PostsGrid Mock
    </div>
    <div
      class="MuiStack-root css-wanzkd-MuiStack-root"
    >
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root MuiPagination-outlined pagination css-1xra8g6-MuiPagination-root"
        data-testid="pagination"
      >
        <ul
          class="MuiPagination-ul css-1c5o7vy-MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-outlinedPrimary Mui-disabled MuiPaginationItem-previousNext css-vs6a7b-MuiButtonBase-root-MuiPaginationItem-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-8uzxat-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateBeforeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-current="true"
              aria-label="page 1"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-outlinedPrimary Mui-selected MuiPaginationItem-page css-vs6a7b-MuiButtonBase-root-MuiPaginationItem-root"
              tabindex="0"
              type="button"
            >
              1
            </button>
          </li>
          <li>
            <button
              aria-label="Go to page 2"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-outlinedPrimary MuiPaginationItem-page css-vs6a7b-MuiButtonBase-root-MuiPaginationItem-root"
              tabindex="0"
              type="button"
            >
              2
            </button>
          </li>
          <li>
            <button
              aria-label="Go to page 3"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-outlinedPrimary MuiPaginationItem-page css-vs6a7b-MuiButtonBase-root-MuiPaginationItem-root"
              tabindex="0"
              type="button"
            >
              3
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular MuiPaginationItem-colorPrimary MuiPaginationItem-outlinedPrimary MuiPaginationItem-previousNext css-vs6a7b-MuiButtonBase-root-MuiPaginationItem-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-8uzxat-MuiSvgIcon-root-MuiPaginationItem-icon"
                data-testid="NavigateNextIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
